// @base-ui-components/react@1.0.0-alpha.5/esm/menu/submenu-trigger/useMenuSubmenuTrigger.js
'use client';

import * as React from 'react';
import { useMenuItem } from '../item/useMenuItem.js';
import { useForkRef } from '../../utils/useForkRef.js';
export function useMenuSubmenuTrigger(parameters) {
  const {
    id,
    highlighted,
    disabled,
    ref: externalRef,
    menuEvents,
    setTriggerElement,
    allowMouseUpTriggerRef,
    typingRef
  } = parameters;
  const {
    getRootProps: getMenuItemProps,
    rootRef: menuItemRef
  } = useMenuItem({
    closeOnClick: false,
    disabled,
    highlighted,
    id,
    menuEvents,
    ref: externalRef,
    allowMouseUpTriggerRef,
    typingRef
  });
  const menuTriggerRef = useForkRef(menuItemRef, setTriggerElement);
  const getRootProps = React.useCallback(externalProps => {
    return {
      ...getMenuItemProps({
        'aria-haspopup': 'menu',
        ...externalProps
      }),
      ref: menuTriggerRef
    };
  }, [getMenuItemProps, menuTriggerRef]);
  return React.useMemo(() => ({
    getRootProps,
    rootRef: menuTriggerRef
  }), [getRootProps, menuTriggerRef]);
}
export let useSubmenuTrigger;
