// @base-ui-components/react@1.0.0-alpha.5/esm/alert-dialog/trigger/AlertDialogTrigger.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { useAlertDialogRootContext } from '../root/AlertDialogRootContext.js';
import { useComponentRenderer } from '../../utils/useComponentRenderer.js';
import { useForkRef } from '../../utils/useForkRef.js';
import { triggerOpenStateMapping } from '../../utils/popupStateMapping.js';

/**
 * A button that opens the alert dialog.
 * Renders a `<button>` element.
 *
 * Documentation: [Base UI Alert Dialog](https://base-ui.com/react/components/alert-dialog)
 */
const AlertDialogTrigger = /*#__PURE__*/React.forwardRef(function AlertDialogTrigger(props, forwardedRef) {
  const {
    render,
    className,
    ...other
  } = props;
  const {
    open,
    setTriggerElement,
    getTriggerProps
  } = useAlertDialogRootContext();
  const state = React.useMemo(() => ({
    open
  }), [open]);
  const mergedRef = useForkRef(forwardedRef, setTriggerElement);
  const {
    renderElement
  } = useComponentRenderer({
    render: render ?? 'button',
    className,
    state,
    propGetter: getTriggerProps,
    extraProps: other,
    customStyleHookMapping: triggerOpenStateMapping,
    ref: mergedRef
  });
  return renderElement();
});
process.env.NODE_ENV !== "production" ? AlertDialogTrigger.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: PropTypes.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
} : void 0;
export { AlertDialogTrigger };
