// @base-ui-components/react@1.0.0-alpha.5/cjs/switch/root/useSwitchRoot.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useSwitchRoot = useSwitchRoot;
var React = _interopRequireWildcard(require("react"));
var _useControlled = require("../../utils/useControlled");
var _useForkRef = require("../../utils/useForkRef");
var _visuallyHidden = require("../../utils/visuallyHidden");
var _mergeReactProps = require("../../utils/mergeReactProps");
var _useEventCallback = require("../../utils/useEventCallback");
var _FieldRootContext = require("../../field/root/FieldRootContext");
var _useBaseUiId = require("../../utils/useBaseUiId");
var _useEnhancedEffect = require("../../utils/useEnhancedEffect");
var _useFieldControlValidation = require("../../field/control/useFieldControlValidation");
var _useField = require("../../field/useField");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function useSwitchRoot(params) {
  const {
    id: idProp,
    checked: checkedProp,
    onCheckedChange: onCheckedChangeProp = () => {},
    defaultChecked,
    name,
    readOnly,
    required,
    disabled = false,
    inputRef: externalInputRef
  } = params;
  const {
    labelId,
    setControlId,
    setTouched,
    setDirty,
    validityData
  } = (0, _FieldRootContext.useFieldRootContext)();
  const {
    getValidationProps,
    getInputValidationProps,
    inputRef: inputValidationRef,
    commitValidation
  } = (0, _useFieldControlValidation.useFieldControlValidation)();
  const onCheckedChange = (0, _useEventCallback.useEventCallback)(onCheckedChangeProp);
  const id = (0, _useBaseUiId.useBaseUiId)(idProp);
  (0, _useEnhancedEffect.useEnhancedEffect)(() => {
    setControlId(id);
    return () => {
      setControlId(undefined);
    };
  }, [id, setControlId]);
  const inputRef = React.useRef(null);
  const handleInputRef = (0, _useForkRef.useForkRef)(inputRef, externalInputRef, inputValidationRef);
  const buttonRef = React.useRef(null);
  const [checked, setCheckedState] = (0, _useControlled.useControlled)({
    controlled: checkedProp,
    default: Boolean(defaultChecked),
    name: 'Switch',
    state: 'checked'
  });
  (0, _useField.useField)({
    id,
    commitValidation,
    value: checked,
    controlRef: buttonRef
  });
  const getButtonProps = React.useCallback((otherProps = {}) => (0, _mergeReactProps.mergeReactProps)(getValidationProps(otherProps), {
    id,
    ref: buttonRef,
    type: 'button',
    role: 'switch',
    disabled,
    'aria-checked': checked,
    'aria-readonly': readOnly,
    'aria-labelledby': labelId,
    onBlur() {
      const element = inputRef.current;
      if (!element) {
        return;
      }
      setTouched(true);
      commitValidation(element.checked);
    },
    onClick(event) {
      if (event.defaultPrevented || readOnly) {
        return;
      }
      inputRef.current?.click();
    }
  }), [getValidationProps, id, disabled, checked, readOnly, labelId, setTouched, commitValidation]);
  const getInputProps = React.useCallback((otherProps = {}) => (0, _mergeReactProps.mergeReactProps)(getInputValidationProps(otherProps), {
    checked,
    disabled,
    name,
    required,
    style: _visuallyHidden.visuallyHidden,
    tabIndex: -1,
    type: 'checkbox',
    'aria-hidden': true,
    ref: handleInputRef,
    onChange(event) {
      // Workaround for https://github.com/facebook/react/issues/9023
      if (event.nativeEvent.defaultPrevented) {
        return;
      }
      const nextChecked = event.target.checked;
      setDirty(nextChecked !== validityData.initialValue);
      setCheckedState(nextChecked);
      onCheckedChange?.(nextChecked, event.nativeEvent);
    }
  }), [getInputValidationProps, checked, disabled, name, required, handleInputRef, setDirty, validityData.initialValue, setCheckedState, onCheckedChange]);
  return React.useMemo(() => ({
    checked,
    getButtonProps,
    getInputProps
  }), [checked, getButtonProps, getInputProps]);
}
