// @base-ui-components/react@1.0.0-alpha.5/esm/number-field/decrement/NumberFieldDecrement.js
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { useNumberFieldRootContext } from '../root/NumberFieldRootContext.js';
import { useComponentRenderer } from '../../utils/useComponentRenderer.js';
/**
 * A stepper button that decreases the field value when clicked.
 * Renders an `<button>` element.
 *
 * Documentation: [Base UI Number Field](https://base-ui.com/react/components/number-field)
 */
const NumberFieldDecrement = /*#__PURE__*/React.forwardRef(function NumberFieldDecrement(props, forwardedRef) {
  const {
    render,
    className,
    ...otherProps
  } = props;
  const {
    getDecrementButtonProps,
    state
  } = useNumberFieldRootContext();
  const {
    renderElement
  } = useComponentRenderer({
    propGetter: getDecrementButtonProps,
    ref: forwardedRef,
    render: render ?? 'button',
    state,
    className,
    extraProps: otherProps
  });
  return renderElement();
});
process.env.NODE_ENV !== "production" ? NumberFieldDecrement.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: PropTypes.node,
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component’s state.
   */
  className: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  /**
   * Allows you to replace the component’s HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
} : void 0;
export { NumberFieldDecrement };
