// @base-ui-components/react@1.0.0-alpha.5/cjs/menu/root/useMenuRoot.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMenuRoot = useMenuRoot;
var React = _interopRequireWildcard(require("react"));
var ReactDOM = _interopRequireWildcard(require("react-dom"));
var _react2 = require("@floating-ui/react");
var _mergeReactProps = require("../../utils/mergeReactProps");
var _useTransitionStatus = require("../../utils/useTransitionStatus");
var _useEventCallback = require("../../utils/useEventCallback");
var _useControlled = require("../../utils/useControlled");
var _constants = require("../../utils/constants");
var _useAfterExitAnimation = require("../../utils/useAfterExitAnimation");
var _useScrollLock = require("../../utils/useScrollLock");
var _translateOpenChangeReason = require("../../utils/translateOpenChangeReason");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const EMPTY_ARRAY = [];
function useMenuRoot(parameters) {
  const {
    open: openParam,
    defaultOpen,
    onOpenChange,
    orientation,
    direction,
    disabled,
    nested,
    closeParentOnEsc,
    loop,
    delay,
    openOnHover,
    onTypingChange,
    modal
  } = parameters;
  const [triggerElement, setTriggerElement] = React.useState(null);
  const [positionerElement, setPositionerElementUnwrapped] = React.useState(null);
  const [instantType, setInstantType] = React.useState();
  const [hoverEnabled, setHoverEnabled] = React.useState(true);
  const [activeIndex, setActiveIndex] = React.useState(null);
  const [openReason, setOpenReason] = React.useState(null);
  const [stickIfOpen, setStickIfOpen] = React.useState(true);
  const popupRef = React.useRef(null);
  const positionerRef = React.useRef(null);
  const stickIfOpenTimeoutRef = React.useRef(-1);
  const [open, setOpenUnwrapped] = (0, _useControlled.useControlled)({
    controlled: openParam,
    default: defaultOpen,
    name: 'useMenuRoot',
    state: 'open'
  });
  const setPositionerElement = React.useCallback(value => {
    positionerRef.current = value;
    setPositionerElementUnwrapped(value);
  }, []);
  const allowMouseUpTriggerRef = React.useRef(false);
  const {
    mounted,
    setMounted,
    transitionStatus
  } = (0, _useTransitionStatus.useTransitionStatus)(open);
  (0, _useScrollLock.useScrollLock)(open && modal, triggerElement);
  const setOpen = (0, _useEventCallback.useEventCallback)((nextOpen, event, reason) => {
    onOpenChange?.(nextOpen, event);
    setOpenUnwrapped(nextOpen);
    if (nextOpen) {
      setOpenReason(reason ?? null);
    }
  });
  (0, _useAfterExitAnimation.useAfterExitAnimation)({
    open,
    animatedElementRef: popupRef,
    onFinished() {
      setMounted(false);
      setOpenReason(null);
      setHoverEnabled(true);
      setStickIfOpen(true);
    }
  });
  const clearStickIfOpenTimeout = (0, _useEventCallback.useEventCallback)(() => {
    clearTimeout(stickIfOpenTimeoutRef.current);
  });
  React.useEffect(() => {
    if (!open) {
      clearStickIfOpenTimeout();
    }
  }, [clearStickIfOpenTimeout, open]);
  React.useEffect(() => {
    return () => {
      clearStickIfOpenTimeout();
    };
  }, [clearStickIfOpenTimeout]);
  const floatingRootContext = (0, _react2.useFloatingRootContext)({
    elements: {
      reference: triggerElement,
      floating: positionerElement
    },
    open,
    onOpenChange(openValue, eventValue, reasonValue) {
      const isHover = reasonValue === 'hover' || reasonValue === 'safe-polygon';
      const isKeyboardClick = reasonValue === 'click' && eventValue.detail === 0;
      const isDismissClose = !openValue && (reasonValue === 'escape-key' || reasonValue == null);
      function changeState() {
        setOpen(openValue, eventValue, (0, _translateOpenChangeReason.translateOpenChangeReason)(reasonValue));
      }
      if (isHover) {
        // Only allow "patient" clicks to close the popover if it's open.
        // If they clicked within 500ms of the popover opening, keep it open.
        clearStickIfOpenTimeout();
        stickIfOpenTimeoutRef.current = window.setTimeout(() => {
          setStickIfOpen(false);
        }, _constants.PATIENT_CLICK_THRESHOLD);
        ReactDOM.flushSync(changeState);
      } else {
        changeState();
      }
      if (isKeyboardClick || isDismissClose) {
        setInstantType(isKeyboardClick ? 'click' : 'dismiss');
      } else {
        setInstantType(undefined);
      }
    }
  });
  const hover = (0, _react2.useHover)(floatingRootContext, {
    enabled: hoverEnabled && openOnHover && !disabled && openReason !== 'click',
    handleClose: (0, _react2.safePolygon)({
      blockPointerEvents: true
    }),
    mouseOnly: true,
    move: false,
    delay: {
      open: delay
    }
  });
  const click = (0, _react2.useClick)(floatingRootContext, {
    enabled: !disabled,
    event: 'mousedown',
    toggle: !nested,
    ignoreMouse: nested,
    stickIfOpen
  });
  const dismiss = (0, _react2.useDismiss)(floatingRootContext, {
    bubbles: closeParentOnEsc && nested,
    outsidePressEvent: 'mousedown'
  });
  const role = (0, _react2.useRole)(floatingRootContext, {
    role: 'menu'
  });
  const itemDomElements = React.useRef([]);
  const itemLabels = React.useRef([]);
  const listNavigation = (0, _react2.useListNavigation)(floatingRootContext, {
    enabled: !disabled,
    listRef: itemDomElements,
    activeIndex,
    nested,
    loop,
    orientation,
    rtl: direction === 'rtl',
    disabledIndices: EMPTY_ARRAY,
    onNavigate: setActiveIndex
  });
  const typeahead = (0, _react2.useTypeahead)(floatingRootContext, {
    listRef: itemLabels,
    activeIndex,
    resetMs: _constants.TYPEAHEAD_RESET_MS,
    onMatch: index => {
      if (open && index !== activeIndex) {
        setActiveIndex(index);
      }
    },
    onTypingChange
  });
  const {
    getReferenceProps,
    getFloatingProps,
    getItemProps
  } = (0, _react2.useInteractions)([hover, click, dismiss, role, listNavigation, typeahead]);
  const getTriggerProps = React.useCallback(externalProps => getReferenceProps((0, _mergeReactProps.mergeReactProps)(externalProps, {
    onMouseEnter() {
      setHoverEnabled(true);
    }
  })), [getReferenceProps]);
  const getPopupProps = React.useCallback(externalProps => getFloatingProps((0, _mergeReactProps.mergeReactProps)(externalProps, {
    onMouseEnter() {
      if (!openOnHover || nested) {
        setHoverEnabled(false);
      }
    },
    onClick() {
      if (openOnHover) {
        setHoverEnabled(false);
      }
    }
  })), [getFloatingProps, openOnHover, nested]);
  return React.useMemo(() => ({
    activeIndex,
    allowMouseUpTriggerRef,
    floatingRootContext,
    getItemProps,
    getPopupProps,
    getTriggerProps,
    itemDomElements,
    itemLabels,
    mounted,
    open,
    popupRef,
    positionerRef,
    setOpen,
    setPositionerElement,
    setTriggerElement,
    transitionStatus,
    openReason,
    instantType
  }), [activeIndex, floatingRootContext, getItemProps, getPopupProps, getTriggerProps, itemDomElements, itemLabels, mounted, open, positionerRef, setOpen, transitionStatus, setPositionerElement, openReason, instantType]);
}
