// @base-ui-components/react@1.0.0-alpha.5/cjs/popover/root/usePopoverRoot.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.usePopoverRoot = usePopoverRoot;
var React = _interopRequireWildcard(require("react"));
var ReactDOM = _interopRequireWildcard(require("react-dom"));
var _react2 = require("@floating-ui/react");
var _useControlled = require("../../utils/useControlled");
var _useEventCallback = require("../../utils/useEventCallback");
var _useTransitionStatus = require("../../utils/useTransitionStatus");
var _constants = require("../utils/constants");
var _mergeReactProps = require("../../utils/mergeReactProps");
var _useOpenInteractionType = require("../../utils/useOpenInteractionType");
var _translateOpenChangeReason = require("../../utils/translateOpenChangeReason");
var _useAfterExitAnimation = require("../../utils/useAfterExitAnimation");
var _constants2 = require("../../utils/constants");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function usePopoverRoot(params) {
  const {
    open: externalOpen,
    onOpenChange: onOpenChangeProp = () => {},
    defaultOpen = false,
    delay,
    closeDelay,
    openOnHover = false
  } = params;
  const delayWithDefault = delay ?? _constants.OPEN_DELAY;
  const closeDelayWithDefault = closeDelay ?? 0;
  const [instantType, setInstantType] = React.useState();
  const [titleId, setTitleId] = React.useState();
  const [descriptionId, setDescriptionId] = React.useState();
  const [triggerElement, setTriggerElement] = React.useState(null);
  const [positionerElement, setPositionerElement] = React.useState(null);
  const [openReason, setOpenReason] = React.useState(null);
  const [clickEnabled, setClickEnabled] = React.useState(true);
  const popupRef = React.useRef(null);
  const clickEnabledTimeoutRef = React.useRef(-1);
  const [open, setOpenUnwrapped] = (0, _useControlled.useControlled)({
    controlled: externalOpen,
    default: defaultOpen,
    name: 'Popover',
    state: 'open'
  });
  if (!open && !clickEnabled) {
    setClickEnabled(true);
  }
  const onOpenChange = (0, _useEventCallback.useEventCallback)(onOpenChangeProp);
  const {
    mounted,
    setMounted,
    transitionStatus
  } = (0, _useTransitionStatus.useTransitionStatus)(open);
  const setOpen = (0, _useEventCallback.useEventCallback)((nextOpen, event, reason) => {
    onOpenChange(nextOpen, event, reason);
    setOpenUnwrapped(nextOpen);
    if (nextOpen) {
      setOpenReason(reason ?? null);
    }
  });
  (0, _useAfterExitAnimation.useAfterExitAnimation)({
    open,
    animatedElementRef: popupRef,
    onFinished() {
      setMounted(false);
      setOpenReason(null);
    }
  });
  React.useEffect(() => {
    return () => {
      clearTimeout(clickEnabledTimeoutRef.current);
    };
  }, []);
  const context = (0, _react2.useFloatingRootContext)({
    elements: {
      reference: triggerElement,
      floating: positionerElement
    },
    open,
    onOpenChange(openValue, eventValue, reasonValue) {
      const isHover = reasonValue === 'hover' || reasonValue === 'safe-polygon';
      const isKeyboardClick = reasonValue === 'click' && eventValue.detail === 0;
      const isDismissClose = !openValue && (reasonValue === 'escape-key' || reasonValue == null);
      function changeState() {
        setOpen(openValue, eventValue, (0, _translateOpenChangeReason.translateOpenChangeReason)(reasonValue));
      }
      if (isHover) {
        // Prevent impatient clicks from unexpectedly closing the popover.
        setClickEnabled(false);
        clearTimeout(clickEnabledTimeoutRef.current);
        clickEnabledTimeoutRef.current = window.setTimeout(() => {
          setClickEnabled(true);
        }, _constants2.PATIENT_CLICK_THRESHOLD);
        ReactDOM.flushSync(changeState);
      } else {
        changeState();
      }
      if (isKeyboardClick || isDismissClose) {
        setInstantType(isKeyboardClick ? 'click' : 'dismiss');
      } else {
        setInstantType(undefined);
      }
    }
  });
  const computedRestMs = delayWithDefault;
  const hover = (0, _react2.useHover)(context, {
    enabled: openOnHover,
    mouseOnly: true,
    move: false,
    handleClose: (0, _react2.safePolygon)(),
    restMs: computedRestMs,
    delay: {
      close: closeDelayWithDefault
    }
  });
  const click = (0, _react2.useClick)(context, {
    enabled: clickEnabled,
    stickIfOpen: false
  });
  const dismiss = (0, _react2.useDismiss)(context);
  const role = (0, _react2.useRole)(context);
  const {
    getReferenceProps,
    getFloatingProps
  } = (0, _react2.useInteractions)([hover, click, dismiss, role]);
  const {
    openMethod,
    triggerProps
  } = (0, _useOpenInteractionType.useOpenInteractionType)(open);
  return React.useMemo(() => ({
    open,
    setOpen,
    mounted,
    setMounted,
    transitionStatus,
    setTriggerElement,
    positionerElement,
    setPositionerElement,
    popupRef,
    titleId,
    setTitleId,
    descriptionId,
    setDescriptionId,
    getRootTriggerProps: externalProps => getReferenceProps((0, _mergeReactProps.mergeReactProps)(externalProps, triggerProps)),
    getRootPopupProps: getFloatingProps,
    floatingRootContext: context,
    instantType,
    openMethod,
    openReason
  }), [mounted, open, setMounted, setOpen, transitionStatus, positionerElement, titleId, descriptionId, getReferenceProps, getFloatingProps, context, instantType, openMethod, triggerProps, openReason]);
}
